### PR TITLE
Fix incorrect memory allocation intervals

### PR DESCRIPTION
**Pull request by brucelwl**
bug fix https://github.com/netty/netty/issues/9549

and pull request https://github.com/netty/netty/pull/9551#event-2618173622

### CODE REVIEW

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java [hunk 1]
```diff
@@ -54,6 +54,26 @@ public void rampUpBeforeReadCompleteWhenLargeDataPending() {
...
+    public void memoryAllocationIntervalsTest() {
+        computingNext(512, 512);
+        computingNext(8192, 1110);
+        computingNext(8192, 1200);
+        computingNext(4096, 1300);
+        computingNext(4096, 1500);
+        computingNext(2048, 1700);
+        computingNext(2048, 1550);
+        computingNext(2048, 2000);
+        computingNext(2048, 1900);
```
Comment (johnou) [inline transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: Can we also assert the guess incrementing after the read bytes increases?

Comment (brucelwl) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: yes, But I think my unit test is correct.

Comment (johnou) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: I'm not saying it's incorrect, but it only asserts in one direction.

Comment (brucelwl) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: Maybe method ``rampUpBeforeReadCompleteWhenLargeDataPending`` contains the assertions you need.

Comment (johnou) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: Point was to test going down and up to ensure the buffer is adaptive.

Comment (normanmaurer) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: @brucelwl can you add the asserts ?

Comment (brucelwl) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: Sorry, I don't have much time to add such code. Maybe you can turn off this pr, but please fix issue https://github.com/netty/netty/issues/9549

Comment (johnou) [reply transport/src/test/java/io/netty/channel/AdaptiveRecvByteBufAllocatorTest.java]: After playing with this a bit I feel the unit test is flawed as it does not take into consideration the index increment and decrement step.

**Comment by johnou** [comment]
> The fix looks correct to me and probably what was originally intended, but would be good to get confirmation from original author (@trustin I think!)

Wouldn't it make more sense to calculate the predicted buffer size using steps and the table instead of asserting random values?

**Comment by njhill** [comment]
@johnou I was referring to the fix itself, I don't disagree with your comments about the unit test... sorry I should have also added that caveat to my approval!

**Comment by trustin** [comment]
TBH I don't remember the reasoning behind this bug (or not) anymore :sweat_smile: The fix seems OK to me though. Would be nice to have a better test case, as commented by @johnou.

**Comment by brucelwl** [comment]
@trustin @johnou  The program I modified meets all the original unit tests. If the unit tests are not perfect, then the original unit tests are not perfect.

**Comment by normanmaurer** [comment]
Let me merge the fix and Improve the unit test as a followup to make progress

**Comment by trustin** [comment]
Thanks @brucelwl for your patience. +1 for moving forward.

**Comment by normanmaurer** [comment]
@brucelwl before I can merge this you need to sign our ICLA: https://netty.io/s/icla

**Comment by brucelwl** [comment]
> @brucelwl before I can merge this you need to sign our ICLA: https://netty.io/s/icla
@normanmaurer I have signed ICLA

**Comment by normanmaurer** [comment]
@brucelwl thanks a lot!

**Pull request merged**